### PR TITLE
Simplify field access for records and record updates

### LINKED ISSUE #34: Simplify field access to records and record updates
**What the rule should do:**

**What problems does it solve:**
Remove redundant record creation/update if immediately followed by a field access

**Example of things the rule would report:**

```elm
{ foo | a = 3 }.a
--> (3)

{ foo | a = f x y }.a
--> (f x y)

{ a = 3 }.a
--> (3)
```

**Example of things the rule would not report:**

```elm
{ foo | b = 3 }.a

{ b = 3 }.a 
```

**Should this be part of the Simplify rule or should it be a new rule?**
Part of the Simplify rule.

**I am looking for:**
- Feedback
- Making a note using an Issue before I forget about this :D I actually plan on writing the PR myself

### PR DESCRIPTION
**Pull request by miniBill**
Closes #34

### CODE REVIEW

File: src/Simplify.elm [hunk 1]
```diff
@@ -1352,10 +1353,64 @@ expressionVisitorHelp node context =
...
+                    in
+                    ( Node.value nfield, nvalue )
+                )
+                setters
+
+        nodeRange =
+            Node.range node
+    in
+    case
+        List.Extra.find (\( setterField, _ ) -> setterField == fieldName) setterValues
```
Comment (jfmengels) [inline src/Simplify.elm:1393]: There is already a `find` function defined in the same module. No need for the dependency.

File: src/Simplify.elm [hunk 2]
```diff
@@ -1352,10 +1353,64 @@ expressionVisitorHelp node context =
...
+            let
+                fieldName =
+                    Node.value field
+            in
+            case Node.value record of
+                Expression.RecordExpr setters ->
+                    onlyErrors (recordAccessChecks node fieldName setters)
+
+                Expression.RecordUpdateExpression _ setters ->
+                    onlyErrors (recordAccessChecks node fieldName setters)
```
Comment (jfmengels) [inline src/Simplify.elm:1366]: You don't need the entire node here, only it's range
```suggestion
                    onlyErrors (recordAccessChecks (Node.range node) fieldName setters)

                Expression.RecordUpdateExpression _ setters ->
                    onlyErrors (recordAccessChecks (Node.range node) fieldName setters)
```

**Review by jfmengels — CHANGES_REQUESTED** [review]
This looks very good. I added a few remarks, but I don't think it's too much work.

File: src/Simplify.elm (hunk 1 above)
Comment (miniBill) [reply src/Simplify.elm:1393]: Ok, let me rewrite it

File: src/Simplify.elm (hunk 2 above)
Comment (miniBill) [reply src/Simplify.elm:1366]: Fair enough!

File: src/Simplify.elm [hunk 3]
```diff
@@ -1352,10 +1353,64 @@ expressionVisitorHelp node context =
...
+recordAccessChecks node fieldName setters =
+    let
+        setterValues =
+            List.map
+                (\setter ->
+                    let
+                        ( nfield, nvalue ) =
+                            Node.value setter
+                    in
+                    ( Node.value nfield, nvalue )
```
[the remark this answers is not in the document]
Comment (miniBill) [reply src/Simplify.elm:1385]: Ooh nice, didn't check whether the constructor was exposed

File: src/Simplify.elm [hunk 4]
```diff
@@ -1352,10 +1353,64 @@ expressionVisitorHelp node context =
...
+            Node.range node
+    in
+    case
+        List.Extra.find (\( setterField, _ ) -> setterField == fieldName) setterValues
+    of
+        Just ( _, setterValue ) ->
+            let
+                setterRange =
+                    Node.range setterValue
+            in
```
[the remark this answers is not in the document]
Comment (miniBill) [reply src/Simplify.elm:1392]: Oh yea, absolutely

**Review by jfmengels — APPROVED** [review]
LGTM, thanks a lot for the new simplification! Very nice addition!

**Pull request merged**